Update return values

jelaniwoods opened the pull request:
Resolves #14 

> At present, the return values of methods are somewhat inconsistent:
> 
> - `generate!` returns a `String`, the content of the response. This is great for being super-quick to see results, especially when writing a Ruby script. But maybe it is abstracting too much — they may benefit from having to `fetch` the content out of a hash? In particular, `response_id` is now important to easily access, especially when storing in a database.
> - `add` returns a `Hash` — perhaps it should return the whole array of messages (similar to `push`)? Or it should return the whole `AI::Chat` object?

---

### Changes


- `generate!` now returns a Hash that contains all the details of the response (some nil, in case it's a background = true request and still queued)
- `add` and its aliases (`system`, `user`, `assistant`)  now return the Array of `messages`

<!-- ELLIPSIS_HIDDEN -->


----

> [!IMPORTANT]
> Update `generate!` to return a `Hash` and `add` to return an `Array` of messages for consistent response handling in `AI::Chat`.
> 
>   - **Behavior**:
>     - `generate!` in `chat.rb` now returns a `Hash` with response details instead of a `String`.
>     - `add` and aliases (`system`, `user`, `assistant`) now return an `Array` of `messages`.
>   - **Examples**:
>     - Updated `01_quick.rb`, `02_core.rb`, and `03_configuration.rb` to reflect new return types.
>     - Modified `04_multimodal.rb` and `05_file_handling_comprehensive.rb` to handle hash responses.
>     - Adjusted `06_structured_output.rb` and `07_structured_output_comprehensive.rb` for structured output handling.
>   - **Documentation**:
>     - Updated `README.md` to show `generate!` returning a `Hash` and `add` returning an `Array` of messages.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=firstdraft%2Fai-chat&utm_source=github&utm_medium=referral)<sup> for 795f0496eefa355e06a981e2208580465e332b82. You can [customize](https://app.ellipsis.dev/firstdraft/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

jelaniwoods committed:
Update return values (#23)

The return values of methods were somewhat inconsistent:
- generate! returned a String, the content of the response. This was great for being super-quick to see results, especially when writing a Ruby script. But it is abstracted a lot — learners may benefit from having to fetch the content out of a hash. Additionally, now that response_id is important to easily access, the String response is less useful.
- add and its aliases returned a Hash — this was inconsistent with our analogy or Array#push

This following updates were introduced in this commit:
- generate! now returns a Hash that contains all of the details of the response.
- add, system, user, assistant now returns the Array of messages.

bpurinton commented on examples/12_image_generation.rb:
```diff
@@ -15,9 +15,10 @@
-puts "User: #{a.user("Draw a simple red circle")}"
-response = a.generate!
-puts "Assistant: #{response}"
+user_message = a.user("Draw a simple red circle")[:content]
```
Getting an error on this line:

```
$ bundle exec ruby examples/12_image_generation.rb 
=== AI::Chat Image Generation Examples ===

Example 1: Basic generation & Response object access
--------------------------------------------------
examples/12_image_generation.rb:18:in `<main>': no implicit conversion of Symbol into Integer (TypeError)

user_message = a.user("Draw a simple red circle")[:content]
                                                  ^^^^^^^^
```

bpurinton replied:
Same error message for the other examples in this file

bpurinton commented on lib/ai/chat.rb:
```diff
@@ -85,11 +84,11 @@ def add(content, role: "user", response: nil, image: nil, images: nil, file: nil
       add(message, role: "system")
     end
-
+    
```
```suggestion

```

minor nit

bpurinton replied:
To remove tab indent (not the blank line, hard to tell from the suggestion diff)

bpurinton commented on lib/ai/chat.rb:
```diff
@@ -85,11 +84,11 @@ def add(content, role: "user", response: nil, image: nil, images: nil, file: nil
       add(message, role: "user", image: image, images: images, file: file, files: files)
     end
-
+    
```
```suggestion

```

bpurinton reviewed (commented):
Just need to double check image generation examples (examples/12), but code changes otherwise LGTM

jelaniwoods committed:
Enable background mode (#24)

This commit adds support for API calls to be made in the background.

A large motivation for this feature is to more easily support deep research models or models that can take a long time to respond.

A message (element from the internal messages Array) now has a status attribute, to represent if the request has completed.

Background requests can be manually polled by method call, which updates the message in the Array messages.
Background requests can "auto-poll" by passing an argument. This renders a spinner and will poll automatically at increasing intervals (with a cap at ~110 seconds) until the response status is completed or the duration exceeds the timeout (default 10 minutes).
